### ISSUE TITLE
`catkin build` causes error about pcl::search::Search

### ISSUE THREAD

**Issue by arwtyxouymz** [ros-planning/moveit_tutorials#258]
Hi,

I use Ubuntu16.04 and ROS Kinetic.
I followed this [Getting Started tutorial](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/getting_started/getting_started.html), and  failed to build this package.

This is the error:
```bash
____________________________________________________________________________________________________________________________________________________________________________________________________________
Errors     << moveit_tutorials:make /home/user/work/examples/ros/ws_moveit/logs/moveit_tutorials/build.make.003.log
CMakeFiles/cylinder_segment.dir/src/cylinder_segment.cpp.o:(.rodata._ZTVN3pcl6search17OrganizedNeighborINS_11PointXYZRGBEEE[_ZTVN3pcl6search17OrganizedNeighborINS_11PointXYZRGBEEE]+0x20): undefined reference to `pcl::search::Search<pcl::PointXYZRGB>::getName[abi:cxx11]() const'
CMakeFiles/cylinder_segment.dir/src/cylinder_segment.cpp.o:(.rodata._ZTVN3pcl6search6SearchINS_11PointXYZRGBEEE[_ZTVN3pcl6search6SearchINS_11PointXYZRGBEEE]+0x20): undefined reference to `pcl::search::Search<pcl::PointXYZRGB>::getName[abi:cxx11]() const'
collect2: error: ld returned 1 exit status
make[2]: *** [/home/user/work/examples/ros/ws_moveit/devel/.private/moveit_tutorials/lib/moveit_tutorials/cylinder_segment] Error 1
make[1]: *** [doc/perception_pipeline/CMakeFiles/cylinder_segment.dir/all] Error 2
make: *** [all] Error 2
cd /home/user/work/examples/ros/ws_moveit/build/moveit_tutorials; catkin build --get-env moveit_tutorials | catkin env -si  /usr/bin/make --jobserver-fds=6,7 -j; cd -
............................................................................................................................................................................................................
```

I checked the pcl_ros and pcl_conversions package written in `CMakeLists.txt` are in my local pc.
How do i fix this?

please help me

**Comment by felixvd** [comment]
Has this been resolved? It looks like a linking problem. I would ask on [ROS Answers](https://answers.ros.org/questions/), and worst case just reinstall from scratch.

**Comment by mlautman** [comment]
If you are having issues with linking I would `catkin clean`, open a new terminal and run `source /opt/ros....` manually, rerun `rosdep install ...`, and see if you get the same issue. Closing for now. If you are still having the same issue, feel free to reopen

**Comment by igorrecioh** [comment]
I am using Ubuntu 18.04 and ROS Melodic.

Same error here:

```
CMakeFiles/cylinder_segment.dir/src/cylinder_segment.cpp.o:(.data.rel.ro._ZTVN3pcl6search17OrganizedNeighborINS_11PointXYZRGBEEE[_ZTVN3pcl6search17OrganizedNeighborINS_11PointXYZRGBEEE]+0x20): undefined reference to `pcl::search::Search<pcl::PointXYZRGB>::getName() const'
CMakeFiles/cylinder_segment.dir/src/cylinder_segment.cpp.o:(.data.rel.ro._ZTVN3pcl6search6SearchINS_11PointXYZRGBEEE[_ZTVN3pcl6search6SearchINS_11PointXYZRGBEEE]+0x20): undefined reference to `pcl::search::Search<pcl::PointXYZRGB>::getName() const'
collect2: error: ld returned 1 exit status
```

**Comment by mlautman** [comment]
Make sure you have the most recent versions of the tutorials and moveit pulled from ros-planning. If you still have trouble, let us know and be sure to tell us more about your workspace. While melodic is being actively developed, it is common for the system installs to fall behind the source code

**Comment by igorrecioh** [comment]
Thank you @mlautman! I will check that I have the latest everything and come back here to update status

**Comment by igorrecioh** [comment]
Hi again!
I solved it by adding 
```
#include <pcl/search/impl/search.hpp>
``` 
in file 
`moveit_tutorials/doc/perception_pipeline/src/cylinder_segment.cpp`

If you prefer, I can make a PR!

**Comment by v4hn** [comment]
@igorrecioh Your fix is likely an unnecessary hack.
You don't use the compiled library functions, but include their sources.

The actual problem seems to be that we find_package `pcl_ros`, but not `PCL` itself, although the code uses library functions.
Could you test whether `find_package(PCL REQUIRED)` + `target_link_libraries(cylinder_segment ${PCL_LIBRARIES})` also fixes the problem on your side?

**Comment by v4hn** [comment]
> `target_link_libraries(cylinder_segment ${PCL_LIBRARIES})`

of course you have to add this line to the appropriated subfolder *after* declaring the build target for the tutorial.

**Comment by aPonza** [comment]
Ubuntu 16.04, ros kinetic, libpcl-dev 1.7.2-14build1, same error as the others

> Could you test whether find_package(PCL REQUIRED) + target_link_libraries(cylinder_segment ${PCL_LIBRARIES}) also fixes the problem on your side?

This didn't fix it for me:

_~/catkin_ws/src/moveit_tutorials/CMakeLists.txt_
```
(...)
    tf2_geometry_msgs
)

find_package(PCL REQUIRED)

find_package(Boost REQUIRED system filesystem date_time thread)
(...)
```

_~/catkin_ws/src/moveit_tutorials/doc/perception_pipeline/CMakeLists.txt_

```
add_executable(cylinder_segment src/cylinder_segment.cpp)
target_link_libraries(cylinder_segment ${catkin_LIBRARIES} ${PCL_LIBRARIES})

add_executable(bag_publisher_maintain_time src/bag_publisher_maintain_time.cpp)
target_link_libraries(bag_publisher_maintain_time ${catkin_LIBRARIES} ${Boost_LIBRARIES})
```

Same error as before but these extra warnings:
```
Warnings   << moveit_tutorials:check ~/catkin_ws/logs/moveit_tutorials/build.check.002.log                                                                                                          
** WARNING ** io features related to openni2 will be disabled
** WARNING ** io features related to pcap will be disabled
** WARNING ** io features related to png will be disabled
** WARNING ** visualization features related to openni2 will be disabled
```

Upon further research it seems it's a problem with v1.7.x as per [this issue](https://github.com/PointCloudLibrary/pcl/issues/2406) where two solutions are listed:
- upgrade to PCL 1.8 (but I'm fully upgraded on xenial, I'd have to build from source);
- `#include <pcl/search/impl/search.hpp>` as a workaround (which works).

Looking further into the warnings there is [issue #119 from ros-perception](https://github.com/ros-perception/perception_pcl/issues/119) and the [ubuntu bugreport](https://bugs.launchpad.net/ubuntu/+source/vtk6/+bug/1573234) not yet fixed in Xenial. I think I'll build PCL from source.

**Comment by progtologist** [comment]
Are you using a non-standard compiler (from a PPA for example)?
Check g++ --version and gcc --version. For kinetic (xenial) you should use gcc version 5.

**Comment by aPonza** [comment]
I worked backwards through what I was doing: I was (and am) using gcc7, and recompiling the whole package with gcc5 doesn't yield the compilation error in OP's post. The alternative, compiling from source, is what I have used up until now since last January. It is giving these warnings, but nothing seems to break:

```
Warnings   << moveit_tutorials:check /home/<user>/catkin_ws/logs/moveit_tutorials/build.check.000.log   
** WARNING ** io features related to pcap will be disabled
** WARNING ** io features related to png will be disabled
** WARNING ** io features related to libusb-1.0 will be disabled
cd /home/<user>/catkin_ws/build/moveit_tutorials; catkin build --get-env moveit_tutorials | catkin env -si  /usr/bin/make cmake_check_build_system; cd -
....................................................................................................
____________________________________________________________________________________________________
Warnings   << moveit_tutorials:make /home/ap/catkin_ws/logs/moveit_tutorials/build.make.001.log     
/usr/bin/ld: warning: libpcl_features.so.1.9, needed by /usr/local/lib/libpcl_segmentation.so, may conflict with libpcl_features.so.1.7
/usr/bin/ld: warning: libpcl_filters.so.1.9, needed by /usr/local/lib/libpcl_segmentation.so, may conflict with libpcl_filters.so.1.7
/usr/bin/ld: warning: libpcl_search.so.1.9, needed by /usr/local/lib/libpcl_segmentation.so, may conflict with libpcl_search.so.1.7
/usr/bin/ld: warning: libpcl_common.so.1.9, needed by /usr/local/lib/libpcl_segmentation.so, may conflict with libpcl_common.so.1.7
cd /home/<user>/catkin_ws/build/moveit_tutorials; catkin build --get-env moveit_tutorials | catkin env -si  /usr/bin/make --jobserver-fds=6,7 -j; cd -
```

**Comment by rconnorlawson** [comment]
I can confirm aPonza's report that using an older version of GCC fixes this issue on Ubuntu 16.04 in ROS Kinetic. 

I was attempting to compile with GCC 8.3.0 and see the same errors as the original poster. Going back to system default (GCC 5.5.0), the error disappears.

**Comment by rhaschke** [comment]
Thanks for the confirmation. Closing this issue.
Solution: Compile with the standard system compiler, gcc 5.
